pkg/store/cache/caching_bucket.go

```diff
@@ -132,35 +131,184 @@ func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailur
 	return cb.WithExpectedErrs(expectedFunc)
 }
 
-var chunksMatcher = regexp.MustCompile(`^.*/chunks/\d+$`)
+func (cb *CachingBucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	cfgName, cfg := cb.cfg.findIterConfig(dir)
+	if cfg == nil {
+		return cb.Bucket.Iter(ctx, dir, f)
+	}
+
+	cb.operationRequests.WithLabelValues(opIter, cfgName).Inc()
+
+	key := cachingKeyIter(dir)
+	data := cfg.cache.Fetch(ctx, []string{key})
+	if data[key] != nil {
+		list, err := cfg.codec.Decode(data[key])
+		if err == nil {
+			cb.operationHits.WithLabelValues(opIter, cfgName).Inc()
+			for _, n := range list {
+				if err := f(n); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+		level.Warn(cb.logger).Log("msg", "failed to decode cached Iter result", "err", err)
+	}
+
+	// Iteration can take a while (esp. since it calls function), and iterTTL is generally low.
+	// We will compute TTL based time when iteration started.
+	iterTime := time.Now()
+	var list []string
+	err := cb.Bucket.Iter(ctx, dir, func(s string) error {
+		list = append(list, s)
+		return f(s)
+	})
+
+	remainingTTL := cfg.ttl - time.Since(iterTime)
+	if err == nil && remainingTTL > 0 {
+		if data, err := cfg.codec.Encode(list); err == nil {
+			cfg.cache.Store(ctx, map[string][]byte{key: data}, remainingTTL)
+		} else {
+			level.Warn(cb.logger).Log("msg", "failed to encode Iter result", "err", err)
+		}
+	}
+	return err
+}
+
+func (cb *CachingBucket) Exists(ctx context.Context, name string) (bool, error) {
+	cfgName, cfg := cb.cfg.findExistConfig(name)
+	if cfg == nil {
+		return cb.Bucket.Exists(ctx, name)
+	}
+
+	cb.operationRequests.WithLabelValues(opExists, cfgName).Inc()
+
+	key := cachingKeyExists(name)
+	hits := cfg.cache.Fetch(ctx, []string{key})
+
+	if ex := hits[key]; ex != nil {
+		exists, err := strconv.ParseBool(string(ex))
+		if err == nil {
+			cb.operationHits.WithLabelValues(opExists, cfgName).Inc()
+			return exists, nil
+		}
+		level.Warn(cb.logger).Log("msg", "unexpected cached 'exists' value", "val", string(ex))
+	}
+
+	existsTime := time.Now()
+	ok, err := cb.Bucket.Exists(ctx, name)
+	if err == nil {
+		storeExistsCacheEntry(ctx, key, ok, existsTime, cfg.cache, cfg.existsTTL, cfg.doesntExistTTL)
+	}
+
+	return ok, err
+}
+
+func storeExistsCacheEntry(ctx context.Context, cachingKey string, exists bool, ts time.Time, cache cache.Cache, existsTTL, doesntExistTTL time.Duration) {
+	var ttl time.Duration
+	if exists {
+		ttl = existsTTL - time.Since(ts)
+	} else {
+		ttl = doesntExistTTL - time.Since(ts)
+	}
+
+	if ttl > 0 {
+		cache.Store(ctx, map[string][]byte{cachingKey: []byte(strconv.FormatBool(exists))}, ttl)
+	}
+}
+
+func (cb *CachingBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
```

Comment:
those files are HUGE!

Reply:
we do this  a lot on compactor

Reply:
> those files are HUGE!

Not sure what you're saying here. If we want to cache file, we need to fetch it into memory first. `Get` is only cached for `meta.json` and deletion mark files. For all other files, `Get` goes directly to bucket, with no caching.

Reply:
`Get` is now streaming.

Reply:
> . For all other files, Get goes directly to bucket, with no caching.

Well, yes, but it's easy to miss. Think about users of this package.

Reply:
Thanks for making it streaming